File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_tree/P543_DiameterOfBinaryTree.java
package com.dsa.leetcode.labuladong.binary_tree;

public class P543_DiameterOfBinaryTree {
    int maxDiameter;

    public int diameterOfBinaryTree(TreeNode root) {
        traverse(root);
        return maxDiameter;
    }

    private int traverse(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = traverse(root.left);
        int right = traverse(root.right);
        maxDiameter = Math.max(maxDiameter, left + right);
        return Math.max(left, right) + 1;
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P543_DiameterOfBinaryTree.py
from typing import Optional

from com.dsa.leetcode.labuladong.binary_tree import TreeNode


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxDiameter = 0

        def maxDepth(node):
            if not node:
                return 0
            left = maxDepth(node.left)
            right = maxDepth(node.right)
            self.maxDiameter = max(self.maxDiameter, left + right)
            return max(left, right) + 1

        maxDepth(root)
        return self.maxDiameter
